rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+}
+```
+
+- instructions
+
+The instructions field is a `Vec` of `Instruction`. An instruction can be
+either an OpCode or some data for that OpCode. For example, the instructions
+`[.., OpCode(Constant), Literal(12), ..]` when evaluated will load the constant
+stored in the `values` `Vec` that is found in position 12 onto the stack.
+
+- constants
+
+A list of constant values found in the program. Since the bytecode only
+contains integers any actual values must be stored here. This also allows
+literals to be deduped.
+
+- targets
+
+A list of paths used in the program, similar to `values`.
+
+- stack
+
+The Vm is a stack based Vm. Every expression that is evaluated pushes the
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- parameter_stack
+
+For function calls, the parameters need to be evaluated and passed to the
+function. The stack is a `Vec` of parameter name and value. We need an OpCode
+that will copy a value from the main stack to the parameter stack tagged with
+the parameter name. The `FunctionArgs` compiler will dump this OpCode after the
+code to evaluate the given arg is dumped.
+
+- error
+
+If an operation errors, it needs to populate this field. Operations such as
+the error assignment (`result, error = thing()`) will need to check if this
+field has been populated and action accordingly.
+
+- instruction_pointer
+
+The instruction pointer points to the next instruction to evaluate.
+
+#### Calling functions
+
+Calling functions in the stdlib will be a case of evaluating each parameter
+with the results pushed onto the parameter stack.
+
+Since VRL allows for named parameters, parameters need to be pushed on to the
+parameter stack _in the order that they are specified in the `Function`
+implementation_.
+
+Parameters are optional and may not be specified. However, to avoid passing
+parameters to the wrong function, an OpCode still needs to be emitted to move
+a placeholder value - `None` - to the parameter stack.
+
+Take the following hypothetical situation:
+
+```coffeescript
+thing(surname: "nong", name: thang(name: "nork"))
+```
+
+Supposing function `thang` also took an optional parameter `surname`, when
+`thang` is called, we will have a `surname` parameter on the parameter stack.
+Without the placeholder, `thang` would thing that `surname` was being passed to
+it. If the placeholder is pushed, it would consume this instead.
+
+We do not want `stdlib` functions to have access to the VM since that risks a
+rogue function destroying the VM's state.
+
+Current `stdlib` functions are composed of a combination of `compile` and
+`resolve` functions. These function will need to be combined into a single
+function `call`.
+
+The `ArgumentList` parameter that is passed into `call` will have access to the
+parameter stack and the parameter list exposed by the `Function`. This can use
+these to return the appropriate values for the `required`, `optional` etc..
+functions.
```

Comment:
I'd love to see a stripped-down example of one of the existing function implementations (f.e. `upcase`), to get a better sense of how this would work, and if this pattern fits all existing use-cases.